runtime/src/chpl-privatization.c

```diff
@@ -25,39 +25,47 @@
 static int64_t chpl_capPrivateObjects = 0;
 static chpl_sync_aux_t privatizationSync;
 
-void** chpl_privateObjects;
+void** chpl_privateObjects = NULL;
 
 void chpl_privatization_init(void) {
     chpl_sync_initAux(&privatizationSync);
 }
 
+static inline int64_t max(int64_t a, int64_t b) {
+  return a > b ? a : b;
+}
+
+// Note that this function can be called in parallel and more notably it can be
+// called with non-monotonic pid's. e.g. this may be called with pid 27, and
+// then pid 2, so it has to ensure that the privatized array has at least pid+1
+// elements. Be __very__ careful if you have to update it.
 void chpl_newPrivatizedClass(void* v, int64_t pid) {
-  // We need to lock around this operation so two calls in rapid succession
-  // that pass the chpl_capPrivateObjects limit don't both try to create a new
-  // array. If they do, one of the calls will be leaked and an invalid pointer
-  // to be placed in the table.
   chpl_sync_lock(&privatizationSync);
 
-  pid += 1;
-  if (pid == 1) {
-    chpl_capPrivateObjects = 8;
-    // "private" means "node-private", so we can use the system allocator.
+  // initialize array to a default size
+  if (chpl_privateObjects == NULL) {
+    chpl_capPrivateObjects = 2*max(pid, 4);
     chpl_privateObjects =
         chpl_mem_allocMany(chpl_capPrivateObjects, sizeof(void *),
                            CHPL_RT_MD_COMM_PRV_OBJ_ARRAY, 0, 0);
   } else {
-    if (pid > chpl_capPrivateObjects) {
+    // if we're out of space, double (or more) the array size
+    if (pid >= chpl_capPrivateObjects) {
       void** tmp;
-      chpl_capPrivateObjects *= 2;
+      int64_t oldCap;
+
+      oldCap = chpl_capPrivateObjects;
+      chpl_capPrivateObjects = 2*pid;
```

Comment:
No, pid of 0 can never result in resizing. In this branch pid is strictly >= chpl_capPrivateObjects, which will always be positive for this resizing branch since the initialize branch will use a minimum size of 0 elements.